<a href="https://colab.research.google.com/github/garyrmeadeTettey/-Machine-Learning-projects/blob/main/NLP_EmailSpam_Detection_script.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
url = 'https://raw.githubusercontent.com/amankharwal/Email-spam-detection/master/emails.csv'
df = pd.read_csv(url, error_bad_lines=False)


/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


In [4]:
df.head(20)
df.tail(20)

text  spam
5708  Subject: re : hi vince  hi jeff ,  no problem ...     0
5709  Subject: agenda for larry thorne ' s presentat...     0
5710  Subject: raptors  here is the most recent vers...     0
5711  Subject: re : faculty lunch  alison ,  i recom...     0
5712  Subject: 2 - survey / information email 5 - 7 ...     0
5713  Subject: promotion  vince , i want to congratu...     0
5714  Subject: re : petronas benchmarking visit  fyi...     0
5715  Subject: request submitted : access request fo...     0
5716  Subject: * special notification * aurora versi...     0
5717  Subject: fwd : update  return - path :  receiv...     0
5718  Subject: altos na gas model  kim , i know you ...     0
5719  Subject: power market research  i came across ...     0
5720  Subject: re : visit to houston  fyi  - - - - -...     0
5721  Subject: ees risk management presentations for...     0
5722  Subject: re : vacation  vince :  i just found ...     0
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

In [5]:
df.shape

(5728, 2)

In [6]:
df.columns

Index(['text', 'spam'], dtype='object')

In [7]:
df.drop_duplicates(inplace=True)
print(df.shape)

(5695, 2)


In [8]:
print(df.isnull().sum())

text    0
spam    0
dtype: int64


In [11]:
!pip install pandas as pd
!pip install numpy as np
!pip install nltk
!pip install string

ERROR: Could not find a version that satisfies the requirement as (from versions: none)
ERROR: No matching distribution found for as
ERROR: Could not find a version that satisfies the requirement as (from versions: none)
ERROR: No matching distribution found for as
ERROR: Could not find a version that satisfies the requirement string (from versions: none)
ERROR: No matching distribution found for string


In [12]:
import numpy as np
import pandas as pd
import nltk
from nltk.corpus import stopwords
import string

In [13]:
# download the stopwords package
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [14]:
def process(text):
    nopunc = [char for char in text if char not in string.punctuation]
    nopunc = ''.join(nopunc)

    clean = [word for word in nopunc.split() if word.lower() not in stopwords.words('english')]
    return clean
# to show the tokenization
df['text'].head().apply(process)

0    [Subject, naturally, irresistible, corporate, ...
1    [Subject, stock, trading, gunslinger, fanny, m...
2    [Subject, unbelievable, new, homes, made, easy...
3    [Subject, 4, color, printing, special, request...
4    [Subject, money, get, software, cds, software,...
Name: text, dtype: object

In [15]:
from sklearn.feature_extraction.text import CountVectorizer
message = CountVectorizer(analyzer=process).fit_transform(df['text'])

split the data into training and testing sets, and then we will use this one row of data for testing to make our prediction later on and test to see if the prediction matches with the actual value

In [16]:
#split the data into 80% training and 20% testing
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(message, df['spam'], test_size=0.20, random_state=0)
# To see the shape of the data
print(message.shape)

(5695, 37229)


create and train the Multinomial Naive Bayes classifier which is suitable for classification with discrete features.

In [17]:
# create and train the Naive Bayes Classifier
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB().fit(xtrain, ytrain)

To see the classifiers prediction and actual values on the data set :

In [18]:
print(classifier.predict(xtrain))
print(ytrain.values)

[0 0 0 ... 0 0 0]
[0 0 0 ... 0 0 0]


evaluating the Naive Bayes classifier and the report, confusion matrix & accuracy score.

In [19]:
# Evaluating the model on the training data set
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
pred = classifier.predict(xtrain)
print(classification_report(ytrain, pred))
print()
print("Confusion Matrix: \n", confusion_matrix(ytrain, pred))
print("Accuracy: \n", accuracy_score(ytrain, pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00      3457
           1       0.99      1.00      0.99      1099

    accuracy                           1.00      4556
   macro avg       0.99      1.00      1.00      4556
weighted avg       1.00      1.00      1.00      4556


Confusion Matrix: 
 [[3445   12]
 [   1 1098]]
Accuracy: 
 0.9971466198419666


test the model on the test data set (xtest &  ytest) by printing the predicted value, and the actual value to see if the model can accurately classify the email text

In [20]:
#print the predictions
print(classifier.predict(xtest))
#print the actual values
print(ytest.values)

[1 0 0 ... 0 0 0]
[1 0 0 ... 0 0 0]


evaluate the model on the test data set :

In [21]:
# Evaluating the model on the training data set
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
pred = classifier.predict(xtest)
print(classification_report(ytest, pred))
print()
print("Confusion Matrix: \n", confusion_matrix(ytest, pred))
print("Accuracy: \n", accuracy_score(ytest, pred))

              precision    recall  f1-score   support

           0       1.00      0.99      0.99       870
           1       0.97      1.00      0.98       269

    accuracy                           0.99      1139
   macro avg       0.98      0.99      0.99      1139
weighted avg       0.99      0.99      0.99      1139


Confusion Matrix: 
 [[862   8]
 [  1 268]]
Accuracy: 
 0.9920983318700615


classifier accurately identified the email messages as spam or not spam with 99.2 % accuracy on the test data

In [22]:
print(classification_report)

<function classification_report at 0x7f09b4171830>


In [39]:
for line in df['text']:
  if ytest.all() != 0:
    
    print("It's a spam message")
  else:
    print("Not a spam message")
    

Streaming output truncated to the last 5000 lines.
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message
Not a spam message